receiver/awsxrayreceiver/receiver.go

```diff
@@ -0,0 +1,206 @@
+type xrayReceiver struct {
+	instanceName string
+	udpSock      socketconn.SocketConn
+	logger       *zap.Logger
+	consumer     consumer.TraceConsumer
+	wg           sync.WaitGroup
+	ctx          context.Context
+	startOnce    sync.Once
+	stopOnce     sync.Once
+}
+
+func newReceiver(config *Config,
+	consumer consumer.TraceConsumer,
+	logger *zap.Logger) (component.TraceReceiver, error) {
+
+	if consumer == nil {
+		return nil, componenterror.ErrNilNextConsumer
+	}
+
+	if config.Transport != transport {
+		return nil, fmt.Errorf(
+			"X-Ray receiver only supports ingesting spans through UDP, provided: %s",
+			config.Transport,
+		)
+	}
+
+	logger.Info("listening on endpoint for X-Ray segments",
+		zap.String(transport, config.Endpoint))
+	sock, err := udp.New(config.Endpoint)
+	if err != nil {
+		return nil, err
+	}
+
+	return &xrayReceiver{
+		instanceName: config.Name(),
+		udpSock:      sock,
+		logger:       logger,
+		consumer:     consumer,
+	}, nil
+}
+
+func (x *xrayReceiver) Start(ctx context.Context, host component.Host) error {
+	// TODO: Might want to pass `host` into read() below to report a fatal error
+
+	var err = componenterror.ErrAlreadyStarted
+	x.startOnce.Do(func() {
+		x.ctx = obsreport.ReceiverContext(ctx, x.instanceName, transport, "")
+		for i := 0; i < maxPollerCount; i++ {
+			x.wg.Add(1)
+			go x.poll()
+		}
+		err = nil
+	})
+	return err
+}
+
+func (x *xrayReceiver) Shutdown(_ context.Context) error {
+	var err = componenterror.ErrAlreadyStopped
+	x.stopOnce.Do(func() {
+		x.udpSock.Close()
+		x.wg.Wait()
+		err = nil
+	})
+	return err
+}
+
+// Reference for this port:
+// https://github.com/aws/aws-xray-daemon/blob/master/cmd/tracing/daemon.go#L257
+func (x *xrayReceiver) read(buf *[]byte) (int, error) {
+	bufVal := *buf
+	rlen, err := x.udpSock.Read(bufVal)
+	if err == nil {
+		return rlen, nil
+	}
+	switch err := err.(type) {
+	case net.Error:
+		if !err.Temporary() {
+			return -1, fmt.Errorf("read from UDP socket: %w", &errIrrecoverable{err})
+		}
+	default:
+		return 0, fmt.Errorf("read from UDP socket: %w", &errRecoverable{err})
+	}
+	return 0, fmt.Errorf("read from UDP socket: %w", &errRecoverable{err})
+}
+
+// this function references the implementation in:
+// https://github.com/aws/aws-xray-daemon/blob/master/cmd/tracing/daemon.go#L275
+// However, it intentionally avoids using a buffer pool and just stick to
+// a local buffer for simplicity and improve concurrency (because in the orignal
+// implementation, the buffer pool is shared between 2 poll() cals executed by
+// 2 goroutines, which requires locking/unlocking a sync.Mutex whenever
+// a buffer is Get()/Return()). Also, the buffer returns by Get()
+// (i.e. https://github.com/aws/aws-xray-daemon/blob/d2331c8c4538e55c237b05306a3cf2e919a41ba9/pkg/bufferpool/bufferpool.go#L28)
+// is the same size as the fallBackBuffer
+// (https://github.com/aws/aws-xray-daemon/blob/master/cmd/tracing/daemon.go#L277)
+func (x *xrayReceiver) poll() {
```

Comment:
Should this be `pollUdp`?

Also, you may consider moving the poll loop into `UDP` and pass back valid segment bodies through a channel so performance of `ConsumeTraces` doesn't affect the poll loop.

Reply:
I think it could also be used to make the shutdown more explicit. IIUC, currently we rely on a socket error when calling socket.close to exit this loop but that's more implicit than it could be. I think normally you would pass a signal through a channel to have the poll loop exit, upon which it signals a waitgroup.